### PR TITLE
[setup_state::02] Made `setup_state` work

### PR DESCRIPTION
**Pull request by kasperski95**
- `tmp_state` is a shallow copied because `deepcopy` froze Protostar

### CODE REVIEW

File: protostar/commands/test/tmp_state.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+from copy import deepcopy
+from types import SimpleNamespace
+
+
+class TmpState(SimpleNamespace):
+    SUPPORTED_TYPES = (int, str, bool)
+
+    def validate(self) -> bool:
```
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/commands/test/tmp_state.py:8]: We coud block mutations here also

File: protostar/commands/test/test_execution_environment.py [hunk 2]
```diff
@@ -45,9 +46,11 @@ class TestExecutionEnvironment:
     def __init__(self, include_paths: List[str]):
         self.starknet = None
         self.test_contract: Optional[StarknetContract] = None
+        self.tmp_state = TmpState()
         self._expected_error: Optional[RevertableException] = None
         self._include_paths = include_paths
         self._test_finish_hooks: Set[Callable[[], None]] = set()
+        self._hijacked_hint_locals: Optional[Dict[str, Any]] = {}
```
Comment (MaksymilianDemitraszek) [inline protostar/commands/test/test_execution_environment.py:53]: Why this is optional if we set it here?

File: protostar/commands/test/test_execution_environment.py [hunk 3]
```diff
@@ -88,6 +92,23 @@ def deploy_in_env(
...
+    async def invoke_setup_state(self, fn_name: str) -> None:
+        await self.invoke_test_case(fn_name)
+        assert self._hijacked_hint_locals is not None
+
+        self.tmp_state = (
+            self._hijacked_hint_locals["tmp_state"]
+            if "tmp_state" in self._hijacked_hint_locals
+            else TmpState()
+        )
+        if not self.tmp_state.validate():
```
Comment (MaksymilianDemitraszek) [inline protostar/commands/test/test_execution_environment.py:104]: This check here is unnecessary, we can do this inside TmpState

File: protostar/commands/test/test_execution_environment.py [hunk 4]
```diff
@@ -150,9 +171,11 @@ def modified_run_from_entrypoint(
             **kwargs,
         ):
             if "hint_locals" in kwargs and kwargs["hint_locals"] is not None:
+                self._hijacked_hint_locals = kwargs["hint_locals"]
                 self._inject_cheats_into_hint_locals(
                     kwargs["hint_locals"], kwargs["hint_locals"]["syscall_handler"]
                 )
+                self._inject_state_into_hint_locals(kwargs["hint_locals"])
```
Comment (MaksymilianDemitraszek) [inline protostar/commands/test/test_execution_environment.py:178]: `tmp_state`

File: protostar/commands/test/test_execution_environment.py [hunk 5]
```diff
@@ -88,6 +92,23 @@ def deploy_in_env(
         )
         return contract
 
+    async def invoke_setup_state(self, fn_name: str) -> None:
+        await self.invoke_test_case(fn_name)
+        assert self._hijacked_hint_locals is not None
+
+        self.tmp_state = (
```
Comment (MaksymilianDemitraszek) [inline protostar/commands/test/test_execution_environment.py:99]: This seems unncessary

File: protostar/commands/test/tmp_state.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/commands/test/tmp_state.py:8]: Would be nice to add checks in __getattr__ if variable has been declared

**Review by MaksymilianDemitraszek — CHANGES_REQUESTED** [review]

File: protostar/commands/test/tmp_state.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
+from copy import deepcopy
+from types import SimpleNamespace
+
+
+class TmpState(SimpleNamespace):
```
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/commands/test/tmp_state.py:5]: It is quite "long lasting"

File: protostar/commands/test/test_execution_environment.py (hunk 2 above)
Comment (MaksymilianDemitraszek) [reply protostar/commands/test/test_execution_environment.py:53]: I am not sure why do we need this at all

File: protostar/commands/test/tmp_state.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/commands/test/tmp_state.py:5]: In `pytest` we have `tmpdir`. I wanted to point out to the user in a similar way that the `state` is not shared between test cases.

File: protostar/commands/test/test_execution_environment.py (hunk 5 above)
Comment (kasperski95) [reply protostar/commands/test/test_execution_environment.py:99]: Yep, `tmp_state` will be updated because it's passed by a reference.

File: protostar/commands/test/tmp_state.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/commands/test/tmp_state.py:5]: Maybe `prearranged_state` ? a bit too wordy...

[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/commands/test/tmp_state.py:5]: `case_state`?

[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/commands/test/tmp_state.py:5]: It's better. I don't know. I'll create a poll on Slack.

[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/commands/test/tmp_state.py:5]: Name of the function: `__setup__`
Name of the object: `context`

File: protostar/commands/test/test_environment_exceptions.py [hunk 7]
```diff
@@ -13,6 +13,15 @@ def __str__(self) -> str:
         return str(super().__repr__())
 
 
+class SimpleReportedException(ReportedException):
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/commands/test/test_environment_exceptions.py:16]: `ReportedException` doesn't parse the message. It prints the whole exception.

File: protostar/commands/test/tmp_state.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/commands/test/tmp_state.py]: Overwriting `getattr` conflicts causes issues during deepcopy.

File: protostar/commands/test/test_execution_environment.py (hunk 2 above)
Comment (kasperski95) [reply protostar/commands/test/test_execution_environment.py]: removed
